// gitbook-styleguide/lib/CountWidget.js
"use strict";

var React = require('react');

var CountWidget = React.createClass({
    displayName: "CountWidget",

    propTypes: {
        left: React.PropTypes.number.isRequired,
        right: React.PropTypes.number.isRequired,
        // min-width for rendered element
        minWidth: React.PropTypes.number
    },

    render: function render() {
        var _props = this.props,
            left = _props.left,
            right = _props.right,
            minWidth = _props.minWidth;

        // Total count

        var total = left + right;

        var countHalfStyle = {};
        if (minWidth) {
            countHalfStyle.minWidth = minWidth / 2 + "px";
        }

        return React.createElement(
            "div",
            { className: "CountWidget" },
            React.createElement(
                "div",
                { className: "CountHalf", style: countHalfStyle },
                React.createElement(
                    "div",
                    { className: "CountValue" },
                    left
                ),
                React.createElement(
                    "span",
                    { className: "CountBar CountBar-Left", style: { width: 100 * left / total + "%" } },
                    React.createElement("div", { className: "CountMeter" })
                )
            ),
            React.createElement(
                "div",
                { className: "CountHalf", style: countHalfStyle },
                React.createElement(
                    "div",
                    { className: "CountValue" },
                    right
                ),
                React.createElement(
                    "span",
                    { className: "CountBar CountBar-Right", style: { width: 100 * right / total + "%" } },
                    React.createElement("div", { className: "CountMeter" })
                )
            )
        );
    }
});

module.exports = CountWidget;
