// gitbook-styleguide/lib/FileInput.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var React = require('react');
var Button = require('./Button');

var FileInput = React.createClass({
    displayName: 'FileInput',

    propTypes: {
        children: React.PropTypes.node,
        onChange: React.PropTypes.func,
        name: React.PropTypes.string,
        accept: React.PropTypes.string,
        disabled: React.PropTypes.bool
    },

    /**
     * Click this button
     */
    click: function click() {
        this.refs.input.click();
    },
    onChange: function onChange(e) {
        if (this.props.onChange) {
            this.props.onChange(e);
        }
    },
    render: function render() {
        return React.createElement(
            Button,
            _extends({}, this.props, { onClick: this.click }),
            React.createElement('input', { type: 'file', ref: 'input', name: this.props.name, style: { display: 'none' }, accept: this.props.accept }),
            this.props.children
        );
    }
});

module.exports = FileInput;
