// gitbook-styleguide/lib/Label.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var React = require('react');
var classNames = require('classnames');

var STYLES = require('./STYLES');

var Label = React.createClass({
    displayName: 'Label',

    propTypes: {
        style: React.PropTypes.oneOf(STYLES),
        className: React.PropTypes.string,
        children: React.PropTypes.node
    },

    getDefaultProps: function getDefaultProps() {
        return {
            style: STYLES[0]
        };
    },
    render: function render() {
        var style = this.props.style;
        var className = classNames('label', 'label-' + style, this.props.className);

        return React.createElement(
            'span',
            { className: className },
            this.props.children
        );
    }
});

/**
 * Create a style for labels
 * @param {String} style
 * @return {React.Component}
 */
function createLabelStyle(style) {
    return React.createClass({
        displayName: Label.displayName + style,
        render: function render() {
            return React.createElement(Label, _extends({}, this.props, { style: style.toLowerCase() }));
        }
    });
}

module.exports = Label;
module.exports.Info = createLabelStyle('Info');
module.exports.Danger = createLabelStyle('Danger');
module.exports.Success = createLabelStyle('Success');
module.exports.Warning = createLabelStyle('Warning');
