// gitbook-styleguide/lib/ProgressBar.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var React = require('react');
var classNames = require('classnames');

var Button = require('./Button');
var SIZES = require('./SIZES');
var STYLES = require('./STYLES');

/**
 * Displays an inline progress bar.
 *
 * @type {ReactClass}
 */
var ProgressBar = React.createClass({
    displayName: 'ProgressBar',

    propTypes: {
        className: React.PropTypes.string,
        // Progress between 0 and max
        value: React.PropTypes.number,
        max: React.PropTypes.number,
        // Size of the bar
        size: React.PropTypes.oneOf(SIZES),
        style: React.PropTypes.oneOf(STYLES)
    },

    getDefaultProps: function getDefaultProps() {
        return {
            value: 0,
            max: 100,
            size: 'md',
            style: 'success'
        };
    },
    render: function render() {
        var _props = this.props,
            size = _props.size,
            style = _props.style,
            value = _props.value,
            max = _props.max;


        var className = classNames('ProgressBar', 'size-' + size, 'style-' + style, this.props.className);

        var percent = value * 100 / max;

        return React.createElement(
            'div',
            { className: className },
            React.createElement('div', { className: 'bar', style: { width: percent + '%' } })
        );
    }
});

/**
 * Display a progress bar in a button like container.
 * Perfect for insertion in a toolbar.
 *
 * @type {ReactClass}
 */
var ProgressBarButton = React.createClass({
    displayName: 'ProgressBarButton',

    propTypes: {
        size: React.PropTypes.oneOf(SIZES),
        className: React.PropTypes.string,
        children: React.PropTypes.node
    },

    render: function render() {
        var _props2 = this.props,
            children = _props2.children,
            className = _props2.className,
            props = _objectWithoutProperties(_props2, ['children', 'className']);

        className = classNames('ProgressBarButton', className);

        return React.createElement(
            Button,
            _extends({ className: className }, props),
            children
        );
    }
});

module.exports = ProgressBar;
module.exports.Button = ProgressBarButton;
