// gitbook-styleguide/lib/ScrollArea.js
"use strict";

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var React = require('react');

/**
 * Component to represent a scrollable content,
 * scroll position can be set as a percent.
 *
 * @type {ReactType}
 */
var ScrollArea = React.createClass({
    displayName: "ScrollArea",

    propTypes: {
        children: React.PropTypes.node,
        // Position of scrolling
        top: React.PropTypes.number,
        // Debouncing interval
        debouncing: React.PropTypes.number,
        // When scrolling
        onScroll: React.PropTypes.func
    },

    getDefaultProps: function getDefaultProps() {
        return {
            debouncing: 5
        };
    },
    onDebouncedScroll: function onDebouncedScroll(e) {
        var onScroll = this.props.onScroll;

        var container = this.refs.scroller;

        if (!container || !onScroll) {
            return;
        }

        var scrollPercentage = container.scrollTop / (container.scrollHeight - container.offsetHeight);
        if (scrollPercentage > 1) {
            scrollPercentage = 1;
        } else if (scrollPercentage < 0) {
            scrollPercentage = 0;
        }

        onScroll(scrollPercentage);
    },
    onScroll: function onScroll(e) {
        var _this = this;

        var debouncing = this.props.debouncing;

        this.cancelDebouncing();
        this.timeout = setTimeout(function () {
            return _this.onDebouncedScroll(e);
        }, debouncing);
    },
    cancelDebouncing: function cancelDebouncing() {
        if (this.timeout) {
            clearTimeout(this.timeout);
        }
    },
    setScroll: function setScroll() {
        var top = this.props.top;

        var container = this.refs.scroller;

        if (container) {
            container.scrollTop = top * (container.scrollHeight - container.offsetHeight);
        }
    },
    componentDidUpdate: function componentDidUpdate(prevProps) {
        if (prevProps.top == this.props.top) {
            return;
        }

        this.setScroll();
    },
    componentDidMount: function componentDidMount() {
        this.setScroll();
    },
    componentWillUnmount: function componentWillUnmount() {
        this.cancelDebouncing();
    },
    render: function render() {
        var _props = this.props,
            children = _props.children,
            props = _objectWithoutProperties(_props, ["children"]);

        delete props.top;
        delete props.debouncing;
        delete props.onScroll;

        return React.createElement(
            "div",
            _extends({
                ref: "scroller",
                className: "ScrollArea",
                onScroll: this.onScroll
            }, props),
            children
        );
    }
});

module.exports = ScrollArea;
