// gitbook-styleguide/lib/Tabs.js
'use strict';

var React = require('react');
var classNames = require('classnames');

/**
 * <Tabs>
 *   <Tabs.Nav>
 *     <Tabs.Item active>Hello</Tabs.Item>
 *     <Tabs.Item>World</Tabs.Item>
 *   </Tabs.Nav>
 *   <Tabs.Content>
 *
 *   </Tabs.Content>
 * </Tabs>
 */

var Tabs = React.createClass({
    displayName: 'Tabs',

    propTypes: {
        children: React.PropTypes.node
    },

    render: function render() {
        return React.createElement(
            'div',
            { className: 'tabs' },
            this.props.children
        );
    }
});

var TabsNav = React.createClass({
    displayName: 'TabsNav',

    propTypes: {
        children: React.PropTypes.node
    },

    render: function render() {
        return React.createElement(
            'div',
            { className: 'tabs-nav' },
            this.props.children
        );
    }
});

var TabsItem = React.createClass({
    displayName: 'TabsItem',

    propTypes: {
        children: React.PropTypes.node,
        active: React.PropTypes.bool,
        href: React.PropTypes.string,
        onClick: React.PropTypes.func
    },

    getDefaultProps: function getDefaultProps() {
        return {
            href: '#'
        };
    },
    onClick: function onClick(event) {
        var onClick = this.props.onClick;


        if (onClick) {
            event.preventDefault();
            onClick();
        }
    },
    render: function render() {
        var _props = this.props,
            children = _props.children,
            href = _props.href,
            active = _props.active;


        return React.createElement(
            'li',
            { className: classNames({ active: active }) },
            React.createElement(
                'a',
                { href: href, onClick: this.onClick },
                children
            )
        );
    }
});

var TabsContent = React.createClass({
    displayName: 'TabsContent',

    propTypes: {
        children: React.PropTypes.node
    },

    render: function render() {
        return React.createElement(
            'div',
            { className: 'tabs-content' },
            this.props.children
        );
    }
});

module.exports = Tabs;
module.exports.Nav = TabsNav;
module.exports.Item = TabsItem;
module.exports.Content = TabsContent;
