// gitbook-styleguide/lib/Tags.js
'use strict';

var React = require('react');
var classNames = require('classnames');
var Icon = require('./Icon');

var Tags = React.createClass({
    displayName: 'Tags',

    propTypes: {
        children: React.PropTypes.node
    },

    render: function render() {
        return React.createElement(
            'ul',
            { className: 'tags' },
            this.props.children
        );
    }
});

var TagItem = React.createClass({
    displayName: 'TagItem',

    propTypes: {
        title: React.PropTypes.string,
        href: React.PropTypes.string,
        active: React.PropTypes.bool,
        onClick: React.PropTypes.func,
        // count and showX are exclusive
        count: React.PropTypes.number,
        showX: React.PropTypes.bool
    },

    getDefaultProps: function getDefaultProps() {
        return {
            href: '#',
            showX: false
        };
    },
    onClick: function onClick(e) {
        var onClick = this.props.onClick;


        if (!onClick) {
            return;
        }

        e.preventDefault();
        onClick();
    },
    render: function render() {
        var _props = this.props,
            active = _props.active,
            count = _props.count,
            showX = _props.showX,
            title = _props.title,
            href = _props.href;


        var className = classNames('tag', {
            active: active
        });

        var inner = '';
        if (showX) {
            inner = React.createElement(
                'span',
                { className: 'extra' },
                ' ',
                React.createElement(Icon, { id: 'x' })
            );
        } else if (typeof count !== 'undefined') {
            inner = React.createElement(
                'span',
                { className: 'extra' },
                ' | ',
                count
            );
        }

        return React.createElement(
            'li',
            { className: className },
            React.createElement(
                'a',
                { href: href, onClick: this.onClick },
                React.createElement(
                    'span',
                    { className: 'name' },
                    title
                ),
                inner
            )
        );
    }
});

module.exports = Tags;
module.exports.Item = TagItem;
